File: app/src/main/java/org/thoughtcrime/securesms/mediasend/VideoEditorFragment.java
package org.thoughtcrime.securesms.mediasend;

import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.fragment.app.Fragment;

import org.signal.core.util.logging.Log;
import org.thoughtcrime.securesms.R;
import org.thoughtcrime.securesms.mms.MediaConstraints;
import org.thoughtcrime.securesms.mms.VideoSlide;
import org.thoughtcrime.securesms.scribbles.VideoEditorHud;
import org.thoughtcrime.securesms.util.Throttler;
import org.thoughtcrime.securesms.video.VideoBitRateCalculator;
import org.thoughtcrime.securesms.video.VideoPlayer;

import java.io.IOException;

public class VideoEditorFragment extends Fragment implements VideoEditorHud.EventListener,
                                                             MediaSendPageFragment {

  private static final String TAG = Log.tag(VideoEditorFragment.class);

  private static final String KEY_URI          = "uri";
  private static final String KEY_MAX_OUTPUT   = "max_output_size";
  private static final String KEY_MAX_SEND     = "max_send_size";
  private static final String KEY_IS_VIDEO_GIF = "is_video_gif";

  private final Throttler videoScanThrottle = new Throttler(150);
  private final Handler   handler           = new Handler(Looper.getMainLooper());

            private Controller     controller;
            private Data           data           = new Data();
            private Uri            uri;
            private boolean        isVideoGif;
            private VideoPlayer    player;
  @Nullable private VideoEditorHud hud;
            private Runnable       updatePosition;
            private boolean        isInEdit;
            private boolean        wasPlayingBeforeEdit;

  public static VideoEditorFragment newInstance(@NonNull Uri uri, long maxCompressedVideoSize, long maxAttachmentSize, boolean isVideoGif) {
    Bundle args = new Bundle();
    args.putParcelable(KEY_URI, uri);
    args.putLong(KEY_MAX_OUTPUT, maxCompressedVideoSize);
    args.putLong(KEY_MAX_SEND, maxAttachmentSize);
    args.putBoolean(KEY_IS_VIDEO_GIF, isVideoGif);

    VideoEditorFragment fragment = new VideoEditorFragment();
    fragment.setArguments(args);
    fragment.setUri(uri);
    return fragment;
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    if (getActivity() instanceof Controller) {
      controller = (Controller) getActivity();
    } else if (getParentFragment() instanceof Controller) {
      controller = (Controller) getParentFragment();
    } else {
      throw new IllegalStateException("Parent must implement Controller interface.");
    }
  }

  @Override
  public @Nullable View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
    return inflater.inflate(R.layout.mediasend_video_fragment, container, false);
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);

    player = view.findViewById(R.id.video_player);

    uri        = requireArguments().getParcelable(KEY_URI);
    isVideoGif = requireArguments().getBoolean(KEY_IS_VIDEO_GIF);

    long       maxOutput  = requireArguments().getLong(KEY_MAX_OUTPUT);
    long       maxSend    = requireArguments().getLong(KEY_MAX_SEND);
    VideoSlide slide      = new VideoSlide(requireContext(), uri, 0, isVideoGif);
    boolean    autoplay   = isVideoGif;

    player.setWindow(requireActivity().getWindow());
    player.setVideoSource(slide, autoplay);

    if (slide.isVideoGif()) {
      player.setPlayerCallback(new VideoPlayer.PlayerCallback() {
        @Override
        public void onPlaying() {
          controller.onPlayerReady();
        }

        @Override
        public void onStopped() {
          // Do nothing.
        }

        @Override
        public void onError() {
          controller.onPlayerError();
        }
      });
      player.hideControls();
      player.loopForever();
    } else if (MediaConstraints.isVideoTranscodeAvailable()) {
      hud = view.findViewById(R.id.video_editor_hud);
      hud.setEventListener(this);
      updateHud(data);
      if (data.durationEdited) {
        player.clip(data.startTimeUs, data.endTimeUs, autoplay);
      }
      try {
        hud.setVideoSource(slide, new VideoBitRateCalculator(maxOutput), maxSend);
        hud.setVisibility(View.VISIBLE);
        startPositionUpdates();
      } catch (IOException e) {
        Log.w(TAG, e);
      }

      player.setOnClickListener(v -> {
        player.pause();
        hud.showPlayButton();
      });

      player.setPlayerCallback(new VideoPlayer.PlayerCallback() {

        @Override
        public void onReady() {
          controller.onPlayerReady();
        }

        @Override
        public void onPlaying() {
          hud.fadePlayButton();
        }

        @Override
        public void onStopped() {
          hud.showPlayButton();
        }

        @Override
        public void onError() {
          controller.onPlayerError();
        }
      });
    }
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();

    if (player != null) {
      player.cleanup();
    }
  }

  @Override
  public void onPause() {
    super.onPause();
    notifyHidden();

    stopPositionUpdates();
  }

  @Override
  public void onResume() {
    super.onResume();
    startPositionUpdates();

    if (player != null && isVideoGif) {
      player.play();
    }
  }

  private void startPositionUpdates() {
    if (hud != null && Build.VERSION.SDK_INT >= 23) {
      stopPositionUpdates();
      updatePosition = new Runnable() {
        @Override
        public void run() {
          hud.setPosition(player.getPlaybackPositionUs());
          handler.postDelayed(this, 100);
        }
      };
      handler.post(updatePosition);
    }
  }

  private void stopPositionUpdates() {
    handler.removeCallbacks(updatePosition);
  }

  @Override
  public void onHiddenChanged(boolean hidden) {
    if (hidden) {
      notifyHidden();
    }
  }

  @Override
  public void setUri(@NonNull Uri uri) {
    this.uri = uri;
  }

  @Override
  public @NonNull Uri getUri() {
    return uri;
  }

  @Override
  public @Nullable View getPlaybackControls() {
    if (hud != null && hud.getVisibility() == View.VISIBLE) return null;
    else if (isVideoGif)                                    return null;
    else if (player != null)                                return player.getControlView();
    else                                                    return null;
  }

  @Override
  public @Nullable Object saveState() {
    return data;
  }

  @Override
  public void restoreState(@NonNull Object state) {
    if (state instanceof Data) {
      data = (Data) state;
      if (Build.VERSION.SDK_INT >= 23) {
        updateHud(data);
      }
    } else {
      Log.w(TAG, "Received a bad saved state. Received class: " + state.getClass().getName());
    }
  }

  @RequiresApi(api = 23)
  private void updateHud(Data data) {
    if (hud != null && data.totalDurationUs > 0 && data.durationEdited) {
      hud.setDurationRange(data.totalDurationUs, data.startTimeUs, data.endTimeUs);
    }
  }

  @Override
  public void notifyHidden() {
    pausePlayback();
  }

  public void pausePlayback() {
    if (player != null) {
      player.pause();
      if (hud != null) {
        hud.showPlayButton();
      }
    }
  }

  @Override
  public void onEditVideoDuration(long totalDurationUs, long startTimeUs, long endTimeUs, boolean fromEdited, boolean editingComplete) {
    controller.onTouchEventsNeeded(!editingComplete);

    if (hud != null) {
      hud.hidePlayButton();
    }

    final long clampedStartTime = Math.max(startTimeUs, 0);

    boolean wasEdited      = data.durationEdited;
    boolean durationEdited = clampedStartTime > 0 || endTimeUs < totalDurationUs;

    data.durationEdited  = durationEdited;
    data.totalDurationUs = totalDurationUs;
    data.startTimeUs     = clampedStartTime;
    data.endTimeUs       = endTimeUs;

    if (editingComplete) {
      isInEdit = false;
      videoScanThrottle.clear();
    } else if (!isInEdit) {
      isInEdit = true;
      wasPlayingBeforeEdit = player.isPlaying();
    }

    videoScanThrottle.publish(() -> {
      player.pause();
      if (!editingComplete) {
        player.removeClip(false);
      }
      player.setPlaybackPosition(fromEdited || editingComplete ? clampedStartTime / 1000 : endTimeUs / 1000);
      if (editingComplete) {
        if (durationEdited) {
          player.clip(clampedStartTime, endTimeUs, wasPlayingBeforeEdit);
        } else {
          player.removeClip(wasPlayingBeforeEdit);
        }

        if (!wasPlayingBeforeEdit) {
          hud.showPlayButton();
        }
      }
    });

    if (!wasEdited && durationEdited) {
      controller.onVideoBeginEdit(uri);
    }

    if (editingComplete) {
      controller.onVideoEndEdit(uri);
    }
  }

  @Override
  public void onPlay() {
    player.play();
  }

  @Override
  public void onSeek(long position, boolean dragComplete) {
    if (dragComplete) {
      videoScanThrottle.clear();
    }

    videoScanThrottle.publish(() -> {
      player.pause();
      player.setPlaybackPosition(position);
    });
  }

  public static class Data {
    boolean durationEdited;
    long    totalDurationUs;
    long    startTimeUs;
    long    endTimeUs;

    public boolean isDurationEdited() {
      return durationEdited;
    }

    public @NonNull Bundle getBundle() {
      Bundle bundle = new Bundle();
      bundle.putByte("EDITED", (byte) (durationEdited ? 1 : 0));
      bundle.putLong("TOTAL", totalDurationUs);
      bundle.putLong("START", startTimeUs);
      bundle.putLong("END", endTimeUs);

      return bundle;
    }

    public static @NonNull Data fromBundle(@NonNull Bundle bundle) {
      Data data = new Data();
      data.durationEdited = bundle.getByte("EDITED") == (byte) 1;
      data.totalDurationUs = bundle.getLong("TOTAL");
      data.startTimeUs = bundle.getLong("START");
      data.endTimeUs = bundle.getLong("END");

      return data;
    }
  }

  public interface Controller {

    void onPlayerReady();

    void onPlayerError();

    void onTouchEventsNeeded(boolean needed);

    void onVideoBeginEdit(@NonNull Uri uri);

    void onVideoEndEdit(@NonNull Uri uri);
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/mediasend/VideoEditorFragment.kt
package org.thoughtcrime.securesms.mediasend

import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.RequiresApi
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import org.signal.core.util.logging.Log
import org.thoughtcrime.securesms.R
import org.thoughtcrime.securesms.mediasend.v2.MediaSelectionViewModel
import org.thoughtcrime.securesms.mediasend.v2.videos.VideoTrimData
import org.thoughtcrime.securesms.mms.MediaConstraints
import org.thoughtcrime.securesms.mms.VideoSlide
import org.thoughtcrime.securesms.scribbles.VideoEditorPlayButtonLayout
import org.thoughtcrime.securesms.util.Throttler
import org.thoughtcrime.securesms.util.visible
import org.thoughtcrime.securesms.video.VideoPlayer
import org.thoughtcrime.securesms.video.VideoPlayer.PlayerCallback
import org.thoughtcrime.securesms.video.videoconverter.VideoThumbnailsRangeSelectorView
import org.thoughtcrime.securesms.video.videoconverter.VideoThumbnailsRangeSelectorView.PositionDragListener
import java.io.IOException
import kotlin.time.Duration.Companion.microseconds
import kotlin.time.Duration.Companion.milliseconds

class VideoEditorFragment : Fragment(), PositionDragListener, MediaSendPageFragment {
  private val sharedViewModel: MediaSelectionViewModel by viewModels(ownerProducer = { requireActivity() })
  private val videoScanThrottle = Throttler(150)
  private val handler = Handler(Looper.getMainLooper())

  private var isVideoGif = false
  private var isInEdit = false
  private var isFocused = false
  private var wasPlayingBeforeEdit = false
  private var maxSend: Long = 0
  private lateinit var uri: Uri
  private lateinit var controller: Controller
  private lateinit var player: VideoPlayer
  private lateinit var hud: VideoEditorPlayButtonLayout
  private lateinit var videoTimeLine: VideoThumbnailsRangeSelectorView

  private val updatePosition = object : Runnable {
    override fun run() {
      if (MediaConstraints.isVideoTranscodeAvailable()) {
        val playbackPosition = player.truePlaybackPosition
        if (playbackPosition >= 0) {
          videoTimeLine.setActualPosition(playbackPosition.milliseconds.inWholeMicroseconds)
          handler.postDelayed(this, 100)
        }
      }
    }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    controller = if (activity is Controller) {
      activity as Controller
    } else if (parentFragment is Controller) {
      parentFragment as Controller
    } else {
      throw IllegalStateException("Parent must implement Controller interface.")
    }
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
    return inflater.inflate(R.layout.mediasend_video_fragment, container, false)
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)

    videoTimeLine = requireActivity().findViewById(R.id.video_timeline)

    player = view.findViewById(R.id.video_player)
    hud = view.findViewById(R.id.video_editor_hud)

    uri = requireArguments().getParcelable(KEY_URI)!!
    isVideoGif = requireArguments().getBoolean(KEY_IS_VIDEO_GIF)
    maxSend = requireArguments().getLong(KEY_MAX_SEND)

    val slide = VideoSlide(requireContext(), uri, 0, isVideoGif)
    player.setWindow(requireActivity().window)
    player.setVideoSource(slide, isVideoGif, TAG)

    hud.visible = !slide.isVideoGif

    if (slide.isVideoGif) {
      player.setPlayerCallback(object : PlayerCallback {
        override fun onPlaying() {
          controller.onPlayerReady()
        }

        override fun onStopped() = Unit

        override fun onError() {
          controller.onPlayerError()
        }
      })
      player.hideControls()
      player.loopForever()
      player.play()
    } else {
      hud.setPlayClickListener {
        player.play()
      }
      player.setOnClickListener {
        player.pause()
        hud.showPlayButton()
      }

      player.setPlayerCallback(object : PlayerCallback {
        override fun onReady() {
          controller.onPlayerReady()
        }

        override fun onPlaying() {
          hud.fadePlayButton()
        }

        override fun onStopped() {
          hud.showPlayButton()
        }

        override fun onError() {
          controller.onPlayerError()
        }
      })
    }

    sharedViewModel.state.observe(viewLifecycleOwner) { incomingState ->
      val focusedUri = incomingState.focusedMedia?.uri
      val currentlyFocused = focusedUri != null && focusedUri == uri
      if (MediaConstraints.isVideoTranscodeAvailable()) {
        if (currentlyFocused) {
          if (isVideoGif) {
            player.play()
          } else {
            if (!isFocused) {
              bindVideoTimeline(incomingState.getOrCreateVideoTrimData(uri))
            } else {
              val videoTrimData = if (focusedUri != null) {
                incomingState.getOrCreateVideoTrimData(focusedUri)
              } else {
                VideoTrimData()
              }
              hud.visible = incomingState.isTouchEnabled && !isVideoGif
              onEditVideoDuration(videoTrimData, incomingState.isTouchEnabled)
            }
          }
        } else {
          stopPositionUpdates()
          player.pause()
        }
      }
      isFocused = currentlyFocused
    }
  }

  @RequiresApi(23)
  private fun bindVideoTimeline(data: VideoTrimData) {
    val autoplay = isVideoGif
    val slide = VideoSlide(requireContext(), uri, 0, autoplay)

    if (data.isDurationEdited) {
      player.clip(data.startTimeUs, data.endTimeUs, autoplay)
    }

    if (slide.hasVideo() && !autoplay) {
      try {
        videoTimeLine.registerPlayerDragListener(this)

        hud.visibility = View.VISIBLE
        startPositionUpdates()
      } catch (e: IOException) {
        Log.w(TAG, e)
      }
    }
  }

  override fun onPositionDrag(position: Long) {
    onSeek(position, false)
  }

  override fun onEndPositionDrag(position: Long) {
    onSeek(position, true)
  }

  override fun onDestroyView() {
    super.onDestroyView()

    player.cleanup()
  }

  override fun onPause() {
    super.onPause()
    notifyHidden()

    stopPositionUpdates()
  }

  override fun onResume() {
    super.onResume()
    startPositionUpdates()

    if (isVideoGif) {
      player.play()
    }
  }

  private fun startPositionUpdates() {
    if (Build.VERSION.SDK_INT >= 23) {
      stopPositionUpdates()
      handler.post(updatePosition)
    }
  }

  private fun stopPositionUpdates() {
    handler.removeCallbacks(updatePosition)
  }

  override fun onHiddenChanged(hidden: Boolean) {
    if (hidden) {
      notifyHidden()
    }
  }

  override fun setUri(uri: Uri) {
    this.uri = uri
  }

  override fun getUri(): Uri {
    return uri
  }

  override fun saveState(): Any = Unit

  override fun restoreState(state: Any) = Unit

  override fun notifyHidden() {
    pausePlayback()
  }

  private fun pausePlayback() {
    player.pause()
    hud.showPlayButton()
  }

  @RequiresApi(23)
  private fun onEditVideoDuration(data: VideoTrimData, editingComplete: Boolean) {
    if (editingComplete) {
      isInEdit = false
      videoScanThrottle.clear()
    } else if (!isInEdit) {
      isInEdit = true
      wasPlayingBeforeEdit = player.isPlaying
    }

    if (wasPlayingBeforeEdit) {
      hud.hidePlayButton()
    }

    videoScanThrottle.publish {
      player.pause()
      if (!editingComplete) {
        player.removeClip(false)
      }
      player.playbackPosition = if (editingComplete) data.startTimeUs / 1000 else data.endTimeUs / 1000
      if (editingComplete) {
        if (data.isDurationEdited) {
          player.clip(data.startTimeUs, data.endTimeUs, wasPlayingBeforeEdit)
        } else {
          player.removeClip(wasPlayingBeforeEdit)
        }

        if (!wasPlayingBeforeEdit) {
          hud.showPlayButton()
        }
      }
    }
  }

  private fun onSeek(position: Long, dragComplete: Boolean) {
    if (dragComplete) {
      videoScanThrottle.clear()
    }

    videoScanThrottle.publish {
      player.pause()
      val milliseconds = position.microseconds.inWholeMilliseconds
      player.playbackPosition = milliseconds
    }
  }

  interface Controller {
    fun onPlayerReady()

    fun onPlayerError()

    fun onTouchEventsNeeded(needed: Boolean)
  }

  companion object {
    private val TAG = Log.tag(VideoEditorFragment::class.java)

    private const val KEY_URI = "uri"
    private const val KEY_MAX_SEND = "max_send_size"
    private const val KEY_IS_VIDEO_GIF = "is_video_gif"

    fun newInstance(uri: Uri, maxAttachmentSize: Long, isVideoGif: Boolean): VideoEditorFragment {
      val args = Bundle()
      args.putParcelable(KEY_URI, uri)
      args.putLong(KEY_MAX_SEND, maxAttachmentSize)
      args.putBoolean(KEY_IS_VIDEO_GIF, isVideoGif)

      val fragment = VideoEditorFragment()
      fragment.arguments = args
      fragment.setUri(uri)
      return fragment
    }
  }
}
